rclpy/rclpy/node.py

```diff
@@ -775,14 +932,20 @@ def set_descriptor(
                 ParameterMsg(name=name, value=alternative_value))
 
         # First try keeping the parameter, then try the alternative one.
-        if (
-            self._apply_descriptor_and_set(current_parameter, descriptor) or
-            self._apply_descriptor_and_set(alternative_parameter, descriptor)
-        ):
-            self._descriptors[name] = descriptor
-            return self.get_parameter(name).get_parameter_value()
-        else:
-            raise InvalidParameterValueException(name)
+        # Don't check for read-only since we are applying a new descriptor now.
```

Comment:
This sounds strange

Reply:
The whole descriptor is applied in a single method, which makes sense to me.

When you set a new descriptor that is read-only, you need to skip that check. You can't modify a parameter that is read-only (nor modify its descriptor), but if the existing descriptor is read-only but the new one is not, that's a valid use-case. Therefore, that check is skipped for new descriptors.

The same applies when declaring a new parameter. The first (and only) time the value is set, you can't check for read-only attribute.

Reply:
IIUC, we can currently change a read-only parameter by changing it's descriptor to `read_only=False` and then updating the parameter. My gut reaction is that the descriptor shouldn't be able to change for a read-only parameter, though I'm okay if we leave the behavior as-is for now and update later if we change our mind.

Reply:
Not exactly; the `read_only` condition of the existing descriptor [is checked](https://github.com/ros2/rclpy/pull/357/files#diff-129fafcf34261ae2f52bd43718815308R922) before attempting to set a new descriptor.

> My gut reaction is that the descriptor shouldn't be able to change for a read-only parameter

That's correct, read-only parameters cannot be modified using this method either.

Reply:
Thanks for the clarification.